### PR TITLE
core(user-flow): audit flow from artifacts json

### PR DESCRIPTION
**Pull request by adamraine**
Steps 4 & 5 of https://github.com/GoogleChrome/lighthouse/issues/13364. I combined them because I needed step 5 to test step 4 🤷 

Lol +900,000 from the artifacts file. It contains the DT logs and traces inline for the equivalent of 3 normal LH runs.

Ref
#11313

### CODE REVIEW

File: lighthouse-core/scripts/update-flow-fixtures.js [hunk 1]
```diff
@@ -45,49 +50,92 @@ async function waitForImagesToLoad(page) {
   }, TIMEOUT);
 }
 
-(async () => {
+/**
+ * @param {LH.Result.MeasureEntry[]} timings
+ */
+function normalizeTimingEntries(timings) {
```
Comment (connorjclark) [inline lighthouse-core/scripts/update-flow-fixtures.js:56]: Let's move this to `assetSaver` under a function named `normalizeArtifacts` or `normalizeArtifactsForTesting` (and use it in `lighthouse-core/scripts/update-report-fixtures.js` too)

**Review by connorjclark — APPROVED** [review]
I expected this to work with `-GA`, but I guess that will come later? If so, please rename the PR title to reflect that this isn't actually loading or saving artifacts to disk during normal operation.

File: lighthouse-core/fraggle-rock/user-flow.js [hunk 2]
```diff
@@ -176,6 +166,69 @@ class UserFlow {
...
+    throw new Error('Need at least one step before getting the result');
+  }
+
+  /** @type {LH.FlowResult['steps']} */
+  const steps = [];
+  for (const gatherStep of gatherSteps) {
+    const {artifacts, name, configContext} = gatherStep;
+
+    let runnerOptions;
+    if ('runnerOptions' in gatherStep) {
```
Comment (connorjclark) [inline lighthouse-core/fraggle-rock/user-flow.js:205]:
```suggestion
    if (gatherStep.runnerOptions) {
```

+ just move `runnerOptions` into the type and make it optional? could maybe rename it to `activeRunnerOptions`.

File: lighthouse-core/fraggle-rock/user-flow.js [hunk 3]
```diff
@@ -176,6 +166,69 @@ class UserFlow {
...
+   * @return {LH.UserFlow.FlowArtifacts}
+   */
+  createArtifactsJson() {
+    /** @type {LH.UserFlow.GatherStep[]} */
+    const stepArtifacts = this._gatherSteps.map(step => ({
+      name: step.name,
+      artifacts: step.artifacts,
+      config: step.config,
+      configContext: step.configContext,
+    }));
```
Comment (connorjclark) [inline lighthouse-core/fraggle-rock/user-flow.js:180]: why map this? could just pass the array directly, or at most do a shorthand shallow clone with `[...this._gatherSteps]`

Comment (adamraine) [reply lighthouse-core/fraggle-rock/user-flow.js:180]: `ActiveGatherStep` contains `runnerOptions` which are not serializable. This was to remove that param.

File: lighthouse-core/fraggle-rock/user-flow.js [hunk 4]
```diff
@@ -176,6 +166,69 @@ class UserFlow {
...
+    if ('runnerOptions' in gatherStep) {
+      // If the gather step is active, we can access the runner options directly.
+      runnerOptions = gatherStep.runnerOptions;
+    } else {
+      // If the gather step is not active, we must recreate the runner options.
+      // Step specific configs take precedence over a config for the entire flow.
+      const configJson = gatherStep.config || flowConfig;
+      const {gatherMode} = artifacts.GatherContext;
+      const {config} = initializeConfig(configJson, {...configContext, gatherMode});
+      runnerOptions = {
```
Comment (connorjclark) [inline lighthouse-core/fraggle-rock/user-flow.js:214]: seems we could just maintained a second property like `_gatherStepRunnerOptions` or something, and keep runner options data there. could be an array or a weakmap (if array, this loop changes to `for (let i ...`, if weakmap can keep for loop as is). I think conceptually the weakmap makes sense here.

File: lighthouse-core/fraggle-rock/user-flow.js (hunk 2 above)
Comment (adamraine) [reply lighthouse-core/fraggle-rock/user-flow.js:205]: Ya that's simpler, I like

File: lighthouse-core/fraggle-rock/user-flow.js (hunk 4 above)
Comment (adamraine) [reply lighthouse-core/fraggle-rock/user-flow.js:209]: I decided to go with the weak map approach here. Keeps `GatherStep` type clear of any extra stuff

**Pull request merged**